### PR TITLE
RemoveResourceWrapperAndJsonSerializableWrapper

### PR DESCRIPTION
**Pull request by xinlian12**
Resolves several comments from the previous JsonSerializableUpdate PR:
https://github.com/Azure/azure-sdk-for-java/pull/10895
1. Remove JsonSerializableWrapper
2. Remove ResourceWrapper
3. CosmosError: remove extending from Resource
4. Rename setAnalyticalStoreTimeToLiveInSeconds inside DocumentCollection, so to be consistent with
CosmosContainerProperties
5. Remove ValidateResource from CosmosAsyncContainer

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CompositePath.java [hunk 1]
```diff
@@ -13,7 +13,9 @@
  * "SELECT * FROM c ORDER BY c.age, c.height", then you need to add "/age" and "/height"
  * as composite paths to your composite index.
  */
-public final class CompositePath extends JsonSerializableWrapper{
+public final class CompositePath{
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CompositePath.java:16]: code style: 
```suggestion
public final class CompositePath {
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java [hunk 2]
```diff
@@ -672,4 +689,22 @@ public String toString() {
     public ObjectNode getPropertyBag() {
         return this.propertyBag;
     }
+
+    <T> boolean containsJsonSerializable(Class<T> c) {
+        return CompositePath.class.isAssignableFrom(c)
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java:694]: You should be able to do class equality here as none of the classes is expected to be subclassed.

```suggestion
CompositePath.class.equals(c)
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ExcludedPath.java [hunk 3]
```diff
@@ -9,7 +9,8 @@
 /**
  * Represents an excluded path of the IndexingPolicy in the Azure Cosmos DB database service.
  */
-public final class ExcludedPath extends JsonSerializableWrapper{
+public final class ExcludedPath{
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ExcludedPath.java:12]: codestyle:
```suggestion
public final class ExcludedPath {
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/SqlParameter.java [hunk 4]
```diff
@@ -70,4 +71,11 @@ public SqlParameter setValue(Object value) {
         this.jsonSerializable.set("value", value);
         return this;
     }
+
+    void populatePropertyBag()
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/SqlParameter.java:75]: code style: open braces on the same line as method definition
```suggestion
 void populatePropertyBag() {
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ModelBridgeInternal.java [hunk 5]
```diff
@@ -502,36 +495,115 @@ public static void setFeedOptionsMaxItemCount(FeedOptions feedOptions, Integer m
...
+            ((CosmosError) t).populatePropertyBag();
+        } else {
+            throw new IllegalArgumentException("populatePropertyBag method does not exists in class " + t.getClass());
         }
-        return resourceWrapper.getResource();
     }
 
-    public static JsonSerializable getJsonSerializableFromIndex(Index index) {
-        return index.getJsonSerializable();
+    public static <T> JsonSerializable invokeGetJsonSerializable(T t) {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ModelBridgeInternal.java:549]: I would just call it `getJsonSerializable(T t)`

Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ModelBridgeInternal.java:549]: can `t` be null?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ModelBridgeInternal.java [hunk 6]
```diff
@@ -502,36 +495,115 @@ public static void setFeedOptionsMaxItemCount(FeedOptions feedOptions, Integer m
...
+        } else if (t instanceof CosmosError) {
+            return ((CosmosError) t).getJsonSerializable();
+        } else {
+            throw new IllegalArgumentException("getJsonSerializable method does not exists in class " + t.getClass());
+        }
     }
 
-    public static JsonSerializable getJsonSerializableFromJsonSerializableWrapper(JsonSerializableWrapper jsonSerializableWrapper) {
-        return jsonSerializableWrapper.getJsonSerializable();
+    public static <T> Resource invokeGetResource(T t) {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ModelBridgeInternal.java:585]:
```suggestion
public static <T> Resource getResource(T t) {
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java [hunk 7]
```diff
@@ -429,10 +439,14 @@ public Double getDouble(String propertyName) {
                     T t = (T) ModelBridgeInternal.instantiateJsonSerializable((ObjectNode) n, c);
                     result.add(t);
 
-                } else if (isJsonSerializableWrapper) {
-                    // JsonSerializableWrapper
-                    T t = (T) ModelBridgeInternal.instantiateJsonSerializableWrapper((ObjectNode) n, c);
-                    result.add(t);
+                } else if (containsJsonSerializable) {
+                    try {
+                        T t = c.getDeclaredConstructor(String.class).newInstance(Utils.toJson(Utils.getSimpleObjectMapper(), (ObjectNode) n));
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java:444]: seems to me this may result in double serialization/deserialization.
can't we just pass ObjectNode to the constructor?

```suggestion
T t = c.getDeclaredConstructor(ObjectNode.class).newInstance((ObjectNode) n));
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java [hunk 8]
```diff
@@ -354,8 +364,8 @@ public Double getDouble(String propertyName) {
                 }
             } else if (JsonSerializable.class.isAssignableFrom(c)) {
                 return (T) ModelBridgeInternal.instantiateJsonSerializable((ObjectNode) jsonObj, c);
-            } else if (JsonSerializableWrapper.class.isAssignableFrom(c)) {
-                return (T) ModelBridgeInternal.instantiateJsonSerializableWrapper((ObjectNode) jsonObj, c);
+            } else if (containsJsonSerializable(c)) {
+                return ModelBridgeInternal.instantiateByJsonString(Utils.toJson(Utils.getSimpleObjectMapper(), (ObjectNode) jsonObj), c);
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java:368]: doesn't this result in double json serialization/deserialization?

can we pass ObjectNode instead? otherwise this will serialize to string and the corresponding class will deserialize string back to ObjectNode.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java (hunk 2 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java]: changed to use equals except Index class.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ModelBridgeInternal.java (hunk 5 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ModelBridgeInternal.java]: It could be. 
One example is in CollectionCrudTest::deleteCollection.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java (hunk 7 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java]: Make sense. Changed to use objectNode and added objectNode related constructor which is missing in some of the classes.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java (hunk 8 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/JsonSerializable.java]: Changed to use ObjectNode related constuctor.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosConflictProperties.java [hunk 9]
```diff
@@ -11,7 +11,7 @@
 /**
  * The type Cosmos conflict properties.
  */
-public final class CosmosConflictProperties extends ResourceWrapper{
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosConflictProperties.java:14]: `getId` is a required method in the Properties classes that we removed.

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDatabaseForTest.java [hunk 10]
```diff
@@ -93,13 +93,13 @@ public static void cleanupStaleTestDatabases(DatabaseManager client) {
                 new SqlQuerySpec("SELECT * FROM c WHERE STARTSWITH(c.id, @PREFIX)", sqlParameterList)).collectList().block();
 
         for (CosmosDatabaseProperties db : dbs) {
-            assertThat(db.getId()).startsWith(CosmosDatabaseForTest.SHARED_DB_ID_PREFIX);
+            assertThat(ModelBridgeInternal.getResource(db).getId()).startsWith(CosmosDatabaseForTest.SHARED_DB_ID_PREFIX);
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDatabaseForTest.java:96]: once `getId()` is added back we should undo this change as getId will be required method in CosmosDatabaseProperties.

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDatabaseTest.java [hunk 11]
```diff
@@ -55,7 +55,7 @@ public void afterClass() {
     @Test(groups = {"emulator"}, timeOut = TIMEOUT)
     public void createDatabase_withPropertiesAndOptions() throws CosmosClientException {
         CosmosDatabaseProperties databaseDefinition = new CosmosDatabaseProperties(CosmosDatabaseForTest.generateId());
-        databases.add(databaseDefinition.getId());
+        databases.add(ModelBridgeInternal.getResource(databaseDefinition).getId());
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDatabaseTest.java:58]: once `getId()` is added back we should undo this change as getId will be required method in CosmosDatabaseProperties.

here and in other places for the properties classes.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosConflictProperties.java (hunk 9 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosConflictProperties.java:14]: please add back

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosDatabaseProperties.java [hunk 12]
```diff
@@ -18,7 +18,7 @@
  * more cosmos items. Since databases are an an administrative resource and the Service Key will be required in
  * order to access and successfully complete any action using the User APIs.
  */
-public final class CosmosDatabaseProperties extends ResourceWrapper{
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosDatabaseProperties.java:21]: `getId` is a required method in the Properties classes that we removed.
please add back here and other places.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosConflictProperties.java (hunk 9 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosConflictProperties.java:14]: should I add back all this method?
getId
setId
getResourceId
setResourceId
getSelfLink
setSelfLink
getTimestamp
setTimestamp
getEtag
setEtag
setAltLink
getAltLink
validateResource

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosConflictProperties.java:14]: added the following change based on the discussion:
add getId(), getResourceId, getEtag(), getTimestamp() to the following classes: CosmosConflictPropertiesCosmosContainerProperties
CosmosDatabaseProperties
CosmosPermissionProperties
CosmosStoredProcedureProperties
CosmosTriggerProperties
osmosUserDefinedFunctionProperties
CosmosUserProperties
ThroughputProperties

setId() is added if it is not existed.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosDatabaseProperties.java (hunk 12 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosDatabaseProperties.java:21]: added back

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDatabaseTest.java (hunk 11 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDatabaseTest.java]: removed.

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDatabaseForTest.java (hunk 10 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosDatabaseForTest.java]: reverted change back.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ModelBridgeInternal.java [hunk 13]
```diff
@@ -502,36 +495,117 @@ public static void setFeedOptionsMaxItemCount(FeedOptions feedOptions, Integer m
...
 
-    public static void populatePropertyBagJsonSerializableWrapper(JsonSerializableWrapper jsonSerializableWrapper) {
-        jsonSerializableWrapper.populatePropertyBag();
-    }
-
-    public static Resource getResourceFromResourceWrapper(ResourceWrapper resourceWrapper) {
-        if (resourceWrapper == null) {
-            return null;
+    public static <T> void populatePropertyBag(T t) {
+        if (t instanceof JsonSerializable) {
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ModelBridgeInternal.java:512]: I checked all the places invoking this populatePropertyBag method, we all pre-checked whether it is null or not. So should be fine here

**Comment by xinlian12** [comment]
Perf test result for readThroughput, writeThroughput and queryInClauseParallel from master branch: 

**Read Throughput Master :** 
2020-05-14 02:13:07,507 [main] INFO com.azure.cosmos.benchmark.AsyncReadBenchmark - [1000000] operations performed in [130] seconds.
5/14/20 2:13:07 AM =============================================================
-- Meters ----------------------------------------------------------------------
#Successful Operations
 count = 1000000
 mean rate = 7657.59 events/second
 1-minute rate = 7531.42 events/second
 5-minute rate = 6691.55 events/second
 15-minute rate = 6362.28 events/second
#Unsuccessful Operations
 count = 0
 mean rate = 0.00 events/second
 1-minute rate = 0.00 events/second
 5-minute rate = 0.00 events/second
 15-minute rate = 0.00 events/second

 
 
**Write Throughput Master :** 

2020-05-14 02:30:36,385 [main] INFO com.azure.cosmos.benchmark.AsyncWriteBenchmark - [1000000] operations performed in [730] seconds.
5/14/20 2:30:36 AM =============================================================
-- Meters ----------------------------------------------------------------------
#Successful Operations
 count = 1000000
 mean rate = 1368.64 events/second
 1-minute rate = 1369.66 events/second
 5-minute rate = 1357.81 events/second
 15-minute rate = 1294.09 events/second
#Unsuccessful Operations
 count = 0
 mean rate = 0.00 events/second
 1-minute rate = 0.00 events/second
 5-minute rate = 0.00 events/second
 15-minute rate = 0.00 events/second

 
**Query In Clause Parallel :** 
2020-05-14 04:49:08,433       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [10000] operations performed in [188] seconds.
5/14/20 4:49:08 AM =============================================================
-- Meters ----------------------------------------------------------------------
#Successful Operations
             count = 10000
         mean rate = 53.16 events/second
     1-minute rate = 53.07 events/second
     5-minute rate = 46.32 events/second
    15-minute rate = 42.77 events/second
#Unsuccessful Operations
             count = 0
         mean rate = 0.00 events/second
     1-minute rate = 0.00 events/second
     5-minute rate = 0.00 events/second
    15-minute rate = 0.00 events/second
-- Timers ----------------------------------------------------------------------
Latency
             count = 10000
         mean rate = 53.16 calls/second
     1-minute rate = 53.06 calls/second
     5-minute rate = 46.32 calls/second
    15-minute rate = 42.77 calls/second
               min = 115.78 milliseconds
               max = 683.09 milliseconds
              mean = 187.71 milliseconds
            stddev = 39.10 milliseconds
            median = 183.68 milliseconds
              75% <= 206.80 milliseconds
              95% <= 248.71 milliseconds
              98% <= 274.79 milliseconds
              99% <= 302.49 milliseconds
            99.9% <= 408.56 milliseconds

**Comment by xinlian12** [comment]
Perf test result for readThroughput, writeThroughput and queryInClauseParallel from this branch:

**Read Throughput this branch:**
2020-05-14 02:27:32,728       [main] INFO  com.azure.cosmos.benchmark.AsyncReadBenchmark - [1000000] operations performed in [128] seconds.
5/14/20 2:27:32 AM =============================================================

-- Meters ----------------------------------------------------------------------
#Successful Operations
             count = 1000000
         mean rate = 7792.76 events/second
     1-minute rate = 7638.45 events/second
     5-minute rate = 6804.52 events/second
    15-minute rate = 6484.82 events/second
#Unsuccessful Operations
             count = 0
         mean rate = 0.00 events/second
     1-minute rate = 0.00 events/second
     5-minute rate = 0.00 events/second
    15-minute rate = 0.00 events/second


**Write Throughput Master :**
2020-05-14 02:57:02,211       [main] INFO  com.azure.cosmos.benchmark.AsyncWriteBenchmark - [1000000] operations performed in [610] seconds.
5/14/20 2:57:02 AM =============================================================

-- Meters ----------------------------------------------------------------------
#Successful Operations
             count = 1000000
         mean rate = 1638.28 events/second
     1-minute rate = 1640.46 events/second
     5-minute rate = 1608.27 events/second
    15-minute rate = 1520.37 events/second
#Unsuccessful Operations
             count = 0
         mean rate = 0.00 events/second
     1-minute rate = 0.00 events/second
     5-minute rate = 0.00 events/second
    15-minute rate = 0.00 events/second

**Query In Clause Parallel :**
2020-05-14 03:49:51,739       [cosmos-rntbd-nio-2-6] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - total pages so far: 10000
2020-05-14 03:49:51,740       [main] INFO  com.azure.cosmos.benchmark.AsyncQueryBenchmark - [10000] operations performed in [212] seconds.
5/14/20 3:49:51 AM =============================================================

-- Meters ----------------------------------------------------------------------
#Successful Operations
             count = 10000
         mean rate = 47.15 events/second
     1-minute rate = 47.19 events/second
     5-minute rate = 47.18 events/second
    15-minute rate = 47.19 events/second
#Unsuccessful Operations
             count = 0
         mean rate = 0.00 events/second
     1-minute rate = 0.00 events/second
     5-minute rate = 0.00 events/second
    15-minute rate = 0.00 events/second

-- Timers ----------------------------------------------------------------------
Latency
             count = 10000
         mean rate = 47.15 calls/second
     1-minute rate = 47.19 calls/second
     5-minute rate = 47.18 calls/second
    15-minute rate = 47.19 calls/second
               min = 93.57 milliseconds
               max = 632.70 milliseconds
              mean = 211.39 milliseconds
            stddev = 73.31 milliseconds
            median = 198.39 milliseconds
              75% <= 259.84 milliseconds
              95% <= 340.24 milliseconds
              98% <= 367.48 milliseconds
              99% <= 394.97 milliseconds
            99.9% <= 472.80 milliseconds

**Comment by mitchdenny** [comment]
Just worked for me ...
